### PR TITLE
Move struct handling from unify to AST

### PR DESCRIPTION
**Pull request by josevalim**
We also improve error messages to list all
available fields and introduce Module.Types.Of
module to hold shared functionality between Expr
and Pattern.

### CODE REVIEW

File: lib/elixir/lib/module/types/expr.ex [hunk 1]
```diff
@@ -142,41 +142,47 @@ defmodule Module.Types.Expr do
...
   # %Struct{map | ...}
-  def of_expr({:%, meta, [module, {:%{}, _, [{:|, _, [map, args]}]}]} = expr, stack, context) do
-    context = Remote.check(module, :__struct__, 0, meta, context)
+  def of_expr({:%, meta, [module, {:%{}, _, [{:|, _, [_, _]}]} = update]} = expr, stack, context) do
     stack = push_expr_stack(expr, stack)
 
-    additional_pairs = [{:required, {:atom, :__struct__}, {:atom, module}}]
-    map_update(map, args, additional_pairs, stack, context)
+    with {:ok, struct, context} <- Of.struct(module, meta, context),
+         {:ok, update, context} <- of_expr(update, stack, context) do
```
Comment (ericmj) [inline lib/elixir/lib/module/types/expr.ex:168]: Nice!

File: lib/elixir/lib/module/types/infer.ex [hunk 2]
```diff
@@ -175,18 +154,15 @@ defmodule Module.Types.Infer do
...
-           |> Enum.uniq()
-           |> simplify_struct() do
+           unify_target_optional(target_optional, source_optional, stack, context) do
+      # Remove duplicate pairs from matching in both left and right directions
+      pairs =
+        (source_required_pairs ++
+           target_required_pairs ++
+           source_optional_pairs ++
+           target_optional_pairs)
+        |> Enum.uniq()
```
Comment (ericmj) [inline lib/elixir/lib/module/types/infer.ex:164]: Remove the pipe?

**Review by ericmj — APPROVED** [review]
Beautiful! 💜

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**